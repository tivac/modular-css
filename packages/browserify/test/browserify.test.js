"use strict";

const browserify = require("browserify");
const from       = require("from2-string");
const shell      = require("shelljs");
const read = require("@modular-css/test-utils/read.js")(__dirname);
const bundle = require("./lib/bundle.js");
const plugin = require("../browserify.js");

describe("/browserify.js", () => {
    // Because these tests keep failing CI...
    jest.setTimeout(20000);
        
    afterAll(() => shell.rm("-rf", "./packages/browserify/test/output/browserify"));

    describe("basic functionality", () => {
        it("should not error if no options are supplied", () => {
            const build = browserify();
            const error = jest.fn();

            build.on("error", error);

            build.plugin(plugin);

            expect(error).not.toHaveBeenCalled();
        });
        
        it("should error if an invalid extension is applied", (done) => {
            const build = browserify();
            
            build.on("error", (err) => {
                expect(err).toMatchSnapshot();

                done();
            });

            build.plugin(plugin, { ext : false });
        });

        it("should error on invalid CSS", (done) => {
            const build = browserify({
                    entries : from("require('./packages/browserify/test/specimens/invalid.css');"),
                });

            let errors = 0;

            build.plugin(plugin);

            build.bundle((err) => {
                ++errors;
                
                expect(err).toBeTruthy();
                
                if(errors === 1) {
                    expect(err.name).toMatch(/SyntaxError|CssSyntaxError/);
                    
                    return false;
                }
                
                return done();
            });
        });

        it("should replace require() calls with the exported identifiers", async () => {
            const build = browserify({
                entries : from("require('./packages/browserify/test/specimens/simple.css');"),
            });
            
            build.plugin(plugin);
            
            const out = await bundle(build);
            
            expect(out.toString()).toMatchSnapshot();
        });

        it("should correctly rewrite urls based on the destination file", async () => {
            const build = browserify({
                entries : from("require('./packages/browserify/test/specimens/relative.css');"),
            });
            
            build.plugin(plugin, {
                css : "./packages/browserify/test/output/browserify/relative.css",
            });
            
            await bundle(build);
            
            expect(read("./browserify/relative.css")).toMatchSnapshot();
        });

        it("should use the specified namer function", async () => {
            const build = browserify({
                entries : from("require('./packages/browserify/test/specimens/keyframes.css');"),
            });
            
            build.plugin(plugin, {
                css   : "./packages/browserify/test/output/browserify/namer-fn.css",
                namer : () => "a",
            });
            
            await bundle(build);
            
            expect(read("./browserify/namer-fn.css")).toMatchSnapshot();
        });

        it("should include all CSS dependencies in output css", async () => {
            const build = browserify({
                entries : from("require('./packages/browserify/test/specimens/start.css');"),
            });
            
            build.plugin(plugin, { css : "./packages/browserify/test/output/browserify/include-css-deps.css" });
            
            const out = await bundle(build);

            expect(out).toMatchSnapshot();
            expect(read("./browserify/include-css-deps.css")).toMatchSnapshot();
        });

        it("should write out the complete exported identifiers when `json` is specified", async () => {
            const build = browserify(from("require('./packages/browserify/test/specimens/multiple.css');"));
            
            build.plugin(plugin, {
                json : "./packages/browserify/test/output/browserify/export-all-identifiers.json",
            });
            
            await bundle(build);
            
            expect(read("./browserify/export-all-identifiers.json")).toMatchSnapshot();
        });

        it("should not include duplicate files in the output multiple times", async () => {
            const build = browserify(
                from("require('./packages/browserify/test/specimens/start.css'); require('./packages/browserify/test/specimens/local.css');")
            );
            
            build.plugin(plugin, { css : "./packages/browserify/test/output/browserify/avoid-duplicates.css" });
            
            const out = await bundle(build);

            expect(out).toMatchSnapshot();
            expect(read("./browserify/avoid-duplicates.css")).toMatchSnapshot();
        });
        
        it("should output an inline source map when the debug option is specified", async () => {
            const build = browserify({
                debug   : true,
                entries : from("require('./packages/browserify/test/specimens/start.css');"),
            });
            
            build.plugin(plugin, { css : "./packages/browserify/test/output/browserify/source-maps.css" });
            
            await bundle(build);
            
            expect(read("./browserify/source-maps.css")).toMatchSnapshot();
        });

        it("should output an external source map when the debug option is specified", async () => {
            const build = browserify({
                debug   : true,
                entries : from("require('./packages/browserify/test/specimens/start.css');"),
            });

            build.plugin(plugin, {
                css : "./packages/browserify/test/output/browserify/source-maps.css",
                map : {
                    inline : false,
                },
            });

            await bundle(build);
            
            expect(read("./browserify/source-maps.css")).toMatchSnapshot();
            expect(read("./browserify/source-maps.css.map")).toMatchSnapshot();
        });
    });
});
