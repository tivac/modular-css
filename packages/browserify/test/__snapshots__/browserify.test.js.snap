// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/browserify.js basic functionality should correctly rewrite urls based on the destination file 1`] = `
"/* packages/browserify/test/specimens/relative.css */
.mcdf61ec72_wooga {
    color: red;
    background: url(\\"../../specimens/folder/to.png\\");
}
"
`;

exports[`/browserify.js basic functionality should error if an invalid extension is applied 1`] = `"Missing or invalid \\"ext\\" option: false"`;

exports[`/browserify.js basic functionality should include all CSS dependencies in output css 1`] = `
"(function(){function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s}return e})()({1:[function(require,module,exports){
require('./packages/browserify/test/specimens/start.css');
},{\\"./packages/browserify/test/specimens/start.css\\":4}],2:[function(require,module,exports){
module.exports = {
    \\"folder\\": \\"mcd19f1319_folder\\"
};
},{}],3:[function(require,module,exports){
module.exports = {
    \\"folder\\": \\"white\\",
    \\"one\\": \\"red\\",
    \\"two\\": \\"blue\\",
    \\"booga\\": \\"mcc4e2a8e1_booga\\",
    \\"looga\\": \\"mcc4e2a8e1_booga mcc4e2a8e1_looga\\"
};
},{\\"./packages/browserify/test/specimens/folder/folder.css\\":2}],4:[function(require,module,exports){
module.exports = {
    \\"one\\": \\"red\\",
    \\"two\\": \\"blue\\",
    \\"folder\\": \\"white\\",
    \\"wooga\\": \\"mcc4e2a8e1_booga mc1c78395f_wooga\\",
    \\"booga\\": \\"mc1c78395f_booga\\",
    \\"tooga\\": \\"mc1c78395f_tooga\\"
};
},{\\"./packages/browserify/test/specimens/folder/folder.css\\":2,\\"./packages/browserify/test/specimens/local.css\\":3}]},{},[1]);
"
`;

exports[`/browserify.js basic functionality should include all CSS dependencies in output css 2`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }
"
`;

exports[`/browserify.js basic functionality should not include duplicate files in the output multiple times 1`] = `
"(function(){function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s}return e})()({1:[function(require,module,exports){
require('./packages/browserify/test/specimens/start.css'); require('./packages/browserify/test/specimens/local.css');
},{\\"./packages/browserify/test/specimens/local.css\\":3,\\"./packages/browserify/test/specimens/start.css\\":4}],2:[function(require,module,exports){
module.exports = {
    \\"folder\\": \\"mcd19f1319_folder\\"
};
},{}],3:[function(require,module,exports){
module.exports = {
    \\"folder\\": \\"white\\",
    \\"one\\": \\"red\\",
    \\"two\\": \\"blue\\",
    \\"booga\\": \\"mcc4e2a8e1_booga\\",
    \\"looga\\": \\"mcc4e2a8e1_booga mcc4e2a8e1_looga\\"
};
},{\\"./packages/browserify/test/specimens/folder/folder.css\\":2}],4:[function(require,module,exports){
module.exports = {
    \\"one\\": \\"red\\",
    \\"two\\": \\"blue\\",
    \\"folder\\": \\"white\\",
    \\"wooga\\": \\"mcc4e2a8e1_booga mc1c78395f_wooga\\",
    \\"booga\\": \\"mc1c78395f_booga\\",
    \\"tooga\\": \\"mc1c78395f_tooga\\"
};
},{\\"./packages/browserify/test/specimens/folder/folder.css\\":2,\\"./packages/browserify/test/specimens/local.css\\":3}]},{},[1]);
"
`;

exports[`/browserify.js basic functionality should not include duplicate files in the output multiple times 2`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }
"
`;

exports[`/browserify.js basic functionality should output an external source map when the debug option is specified 1`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }

/*# sourceMappingURL=source-maps.css.map */"
`;

exports[`/browserify.js basic functionality should output an external source map when the debug option is specified 2`] = `"{\\"version\\":3,\\"sources\\":[\\"../../specimens/folder/folder.css\\",\\"../../specimens/local.css\\",\\"../../specimens/start.css\\"],\\"names\\":[],\\"mappings\\":\\"AAAA,0DAAC;AACD,qBAAU,YAAY,EAAE;ACDxB,kDAAC;AAID,oBAAS,kBAAkB,EAAE;ACJ7B,kDAAC;AAED,oBDFA,WAAgB,CAChB,iBAAiB,ECCsB;AACvC,oBFHA,wBAAqB,EEGe\\",\\"file\\":\\"source-maps.css\\",\\"sourcesContent\\":[\\"@value folder: white;\\\\n.folder { margin: 2px; }\\\\n\\",\\"@value one: red;\\\\n@value two: blue;\\\\n@value folder from \\\\\\"./folder/folder.css\\\\\\";\\\\n\\\\n.booga { background: green; }\\\\n.looga { composes: booga; }\\\\n\\",\\"@value one, two, folder from \\\\\\"./local.css\\\\\\";\\\\n.wooga { composes: booga from \\\\\\"./local.css\\\\\\"; }\\\\n.booga { color: one; background: two; }\\\\n.tooga { border: 1px solid folder; }\\\\n\\"]}"`;

exports[`/browserify.js basic functionality should output an inline source map when the debug option is specified 1`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NwZWNpbWVucy9mb2xkZXIvZm9sZGVyLmNzcyIsIi4uLy4uL3NwZWNpbWVucy9sb2NhbC5jc3MiLCIuLi8uLi9zcGVjaW1lbnMvc3RhcnQuY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLDBEQUFDO0FBQ0QscUJBQVUsWUFBWSxFQUFFO0FDRHhCLGtEQUFDO0FBSUQsb0JBQVMsa0JBQWtCLEVBQUU7QUNKN0Isa0RBQUM7QUFFRCxvQkRGQSxXQUFnQixDQUNoQixpQkFBaUIsRUNDc0I7QUFDdkMsb0JGSEEsd0JBQXFCLEVFR2UiLCJmaWxlIjoic291cmNlLW1hcHMuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiQHZhbHVlIGZvbGRlcjogd2hpdGU7XG4uZm9sZGVyIHsgbWFyZ2luOiAycHg7IH1cbiIsIkB2YWx1ZSBvbmU6IHJlZDtcbkB2YWx1ZSB0d286IGJsdWU7XG5AdmFsdWUgZm9sZGVyIGZyb20gXCIuL2ZvbGRlci9mb2xkZXIuY3NzXCI7XG5cbi5ib29nYSB7IGJhY2tncm91bmQ6IGdyZWVuOyB9XG4ubG9vZ2EgeyBjb21wb3NlczogYm9vZ2E7IH1cbiIsIkB2YWx1ZSBvbmUsIHR3bywgZm9sZGVyIGZyb20gXCIuL2xvY2FsLmNzc1wiO1xuLndvb2dhIHsgY29tcG9zZXM6IGJvb2dhIGZyb20gXCIuL2xvY2FsLmNzc1wiOyB9XG4uYm9vZ2EgeyBjb2xvcjogb25lOyBiYWNrZ3JvdW5kOiB0d287IH1cbi50b29nYSB7IGJvcmRlcjogMXB4IHNvbGlkIGZvbGRlcjsgfVxuIl19 */"
`;

exports[`/browserify.js basic functionality should replace require() calls with the exported identifiers 1`] = `
"(function(){function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s}return e})()({1:[function(require,module,exports){
require('./packages/browserify/test/specimens/simple.css');
},{\\"./packages/browserify/test/specimens/simple.css\\":2}],2:[function(require,module,exports){
module.exports = {
    \\"wooga\\": \\"mc459e927d_wooga\\"
};
},{}]},{},[1]);
"
`;

exports[`/browserify.js basic functionality should use the specified namer function 1`] = `
"/* packages/browserify/test/specimens/keyframes.css */
@keyframes a {
      0% { opacity: 0; }
    100% { opacity: 1; }
}

.a {
    animation: a 2s;
}

@-webkit-keyframes a {
    from { background-position: 0px; }
    to { background-position: -3968px;}
}

.a {
    -webkit-animation: a 2.5s steps(62) infinite;
}
"
`;

exports[`/browserify.js basic functionality should write out the complete exported identifiers when \`json\` is specified 1`] = `
"{
    \\"packages/browserify/test/specimens/multiple.css\\": {
        \\"fooga\\": \\"mc9a6b8d68_fooga\\",
        \\"wooga\\": \\"mc9a6b8d68_fooga mc9a6b8d68_wooga\\"
    }
}"
`;
